### PR TITLE
Test that even non-enforced keys get grouped properly

### PR DESCRIPTION
**Pull request by bkirwi**
### Motivation

  * This PR fixes a bug reported by a prospect on Slack.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

Sorry, @morsapaes, but this one will require some wordsmithing. Here's a quick braindump for the note though.

> Fix a bug in sinks that could cause updates to appear as two separate records instead of consolidated into a single update record. For `ENVELOPE UPSERT`, as a deletion tombstone followed by a record with the new value, and for `ENVELOPE DEBEZIUM`, as a `{"before": "OLDVALUE", "after": null}` record followed by a `{"before": null, "after": "NEWVALUE"}` record. The bug occurred when updates for multiple keys occurred at the same timestamp.

### CODE REVIEW

**Comment by benesch** [comment]
I pushed up three additional commits here:

  * An adjustment to the test that reproduces the non-consolidation bug.
  * Frank's fix from #15757, which fixes the bug.
  * A larger refactor to `combine_at_timestamp` that I think improves clarity.

**Comment by benesch** [comment]
Also: pretty nervous that we didn't have any test coverage! The teeny little addition to tickle this bug doesn't give me a ton of confidence. We probably want to throw @philip-stoev at this.

**Comment by philip-stoev** [comment]
Please let me know when this is ready for testing.

**Comment by benesch** [comment]
> Please let me know when this is ready for testing.

This is ready for testing! Let me know if you need more of an explainer than in the test case.

**Comment by benesch** [comment]
@frankmcsherry @bkirwi we could choose to exclude my last refactoring commit, if you don't think it is an improvement in clarity. (I suspect it is also a minor perf improvement, but I didn't measure.) It should be behaviorally equivalent to the code after Frank's fix.

**Review by frankmcsherry — APPROVED** [review]
LGTM! I personally like the refactor, so .. works for me!

File: test/testdrive/kafka-avro-upsert-sinks.td [hunk 1]
```diff
@@ -279,6 +279,27 @@ $ kafka-create-topic topic=non-keyed-input
...
+$ kafka-ingest format=avro topic=non-keyed-input schema=${schema}
+{"array":[{"data":{"a":1,"b":1},"time":1,"diff":1}]}
+{"array":[{"data":{"a":2,"b":1},"time":1,"diff":1}]}
+{"array":[{"data":{"a":1,"b":1},"time":2,"diff":-1}]}
+{"array":[{"data":{"a":2,"b":1},"time":2,"diff":-1}]}
+{"array":[{"data":{"a":1,"b":2},"time":2,"diff":1}]}
+{"array":[{"data":{"a":2,"b":2},"time":2,"diff":1}]}
+{"array":[{"data":{"a":1,"b":2},"time":3,"diff":-1}]}
+{"array":[{"data":{"a":2,"b":2},"time":3,"diff":-1}]}
+{"com.materialize.cdc.progress":{"lower":[0],"upper":[10],"counts":[{"time":1,"count":1}, {"time":2,"count":2}, {"time":3,"count":1}]}}
```
Comment (philip-stoev) [inline test/testdrive/kafka-avro-upsert-sinks.td:292]: How come the sum of the `counts` does not match the number of messages ingested not the difference between upper and lower?

**Review by philip-stoev — APPROVED** [review]
Unfortunately due to https://github.com/MaterializeInc/materialize/issues/10406 all interesting interleavings are invalid and cause a panic in timely. So I was unable to test much more than the test you provided. I am going to paste my test in this PR in case you would like to merge it -- I am not pushing it myself as this PR has one too many authors and I do not want to cause a merge conflict for whoever is merging it.

**Comment by philip-stoev** [comment]
```
# Copyright Materialize, Inc. and contributors. All rights reserved.
#
# Use of this software is governed by the Business Source License
# included in the LICENSE file at the root of this repository.
#
# As of the Change Date specified in that file, in accordance with
# the Business Source License, use of this software will be governed
# by the Apache License, Version 2.0.

#
# This test creates a single timestamp with multiple events in it
#

> CREATE CONNECTION kafka_conn
  TO KAFKA (BROKER '${testdrive.kafka-addr}');

> CREATE CONNECTION IF NOT EXISTS csr_conn TO CONFLUENT SCHEMA REGISTRY (
    URL '${testdrive.schema-registry-url}'
  );

$ set schema=[
  {
    "type": "array",
    "items": {
      "type": "record",
      "name": "update",
      "namespace": "com.materialize.cdc",
      "fields": [
        {
          "name": "data",
          "type": {
            "type": "record",
            "name": "data",
            "fields": [
              {"name": "a", "type": "long"},
              {"name": "b", "type": "long"}
            ]
          }
        },
        {
          "name": "time",
          "type": "long"
        },
        {
          "name": "diff",
          "type": "long"
        }
      ]
    }
  },
  {
    "type": "record",
    "name": "progress",
    "namespace": "com.materialize.cdc",
    "fields": [
      {
        "name": "lower",
        "type": {
          "type": "array",
          "items": "long"
        }
      },
      {
        "name": "upper",
        "type": {
          "type": "array",
          "items": "long"
        }
      },
      {
        "name": "counts",
        "type": {
          "type": "array",
          "items": {
            "type": "record",
            "name": "counts",
            "fields": [
              {
                "name": "time",
                "type": "long"
              },
              {
                "name": "count",
                "type": "long"
              }
            ]
          }
        }
      }
    ]
  }
  ]


#
# Insert some rows and then delete, upsert and insert even more in the same timestamp
#

$ kafka-create-topic topic=topic1
> CREATE SOURCE source1
  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-topic1-${testdrive.seed}')
  FORMAT AVRO USING SCHEMA '${schema}' ENVELOPE MATERIALIZE
> CREATE SINK sink1 FROM source1
  INTO KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-sink1-${testdrive.seed}') KEY (a) NOT ENFORCED
  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn ENVELOPE UPSERT

# We start with 10 records

$ kafka-ingest format=avro topic=topic1 schema=${schema}
{"array":[{"data":{"a":10,"b":1},"time":1,"diff":1}]}
{"array":[{"data":{"a":9,"b":1},"time":1,"diff":1}]}
{"array":[{"data":{"a":8,"b":1},"time":1,"diff":1}]}
{"array":[{"data":{"a":7,"b":1},"time":1,"diff":1}]}
{"array":[{"data":{"a":6,"b":1},"time":1,"diff":1}]}
{"array":[{"data":{"a":5,"b":1},"time":1,"diff":1}]}
{"array":[{"data":{"a":4,"b":1},"time":1,"diff":1}]}
{"array":[{"data":{"a":2,"b":1},"time":1,"diff":1}]}
{"array":[{"data":{"a":3,"b":1},"time":1,"diff":1}]}
{"array":[{"data":{"a":1,"b":1},"time":1,"diff":1}]}

# All of the below happens in a single timestamp "time":2

# Delete 2 records
$ kafka-ingest format=avro topic=topic1 schema=${schema}
{"array":[{"data":{"a":7,"b":1},"time":2,"diff":-1}]}
{"array":[{"data":{"a":3,"b":1},"time":2,"diff":-1}]}

# Upsert 2 records
$ kafka-ingest format=avro topic=topic1 schema=${schema}
{"array":[{"data":{"a":8,"b":1},"time":2,"diff":-1}]}
{"array":[{"data":{"a":2,"b":1},"time":2,"diff":-1}]}
{"array":[{"data":{"a":8,"b":8},"time":2,"diff":1}]}
{"array":[{"data":{"a":2,"b":2},"time":2,"diff":1}]}

# Insert 2 records
$ kafka-ingest format=avro topic=topic1 schema=${schema}
{"array":[{"data":{"a":0,"b":0},"time":2,"diff":1}]}
{"array":[{"data":{"a":15,"b":15},"time":2,"diff":1}]}

# Emit the progress
$ kafka-ingest format=avro topic=topic1 schema=${schema}
{"com.materialize.cdc.progress":{"lower":[0],"upper":[10],"counts":[{"time":1,"count":10}, {"time":2,"count":8}]}}

# Because of sort-messages=true, the deletions must be listed first
$ kafka-verify-data headers=materialize-timestamp format=avro topic=testdrive-sink1-${testdrive.seed}
1	{"a": 1} {"a": 1, "b": 1}
1	{"a": 2} {"a": 2, "b": 1}
1	{"a": 3} {"a": 3, "b": 1}
1	{"a": 4} {"a": 4, "b": 1}
1	{"a": 5} {"a": 5, "b": 1}
1	{"a": 6} {"a": 6, "b": 1}
1	{"a": 7} {"a": 7, "b": 1}
1	{"a": 8} {"a": 8, "b": 1}
1	{"a": 9} {"a": 9, "b": 1}
1	{"a": 10} {"a": 10, "b": 1}
2	{"a": 0} {"a": 0, "b": 0}
2	{"a": 2} {"a": 2, "b": 2}
2	{"a": 3}
2	{"a": 7}
2	{"a": 8} {"a": 8, "b": 8}
2	{"a": 15} {"a": 15, "b": 15}
```

**Comment by benesch** [comment]
Thanks, @philip-stoev! Marking as a release blocker so that we don't forget to merge this before the release tomorrow.

File: src/interchange/src/envelopes.rs [hunk 2]
```diff
@@ -43,72 +48,66 @@ where
...
-                            .into_iter()
-                            .group_by(|(t, k, _diff, _row)| (*t, k.clone()))
-                        {
-                            let mut out = vec![];
-                            let elts: Vec<(G::Timestamp, Option<&Row>, Diff, &Row)> =
-                                group.collect();
-                            // Step (2) above
-                            let pos_idx = elts
-                                .binary_search_by(|(_t, _k, diff, _row)| diff.cmp(&0))
-                                .expect_err("there should be no zero-multiplicity entries");
```
Comment (philip-stoev) [inline src/interchange/src/envelopes.rs:86]: I curiously, I was unable to trigger this even in the old code -- timely will panic with some other panic text if there is a diff = 0 or if a record is both inserted and deleted at the same timestamp.

File: test/testdrive/kafka-avro-upsert-sinks.td (hunk 1 above)
Comment (frankmcsherry) [reply test/testdrive/kafka-avro-upsert-sinks.td:292]: I think the counts just need to be doubled. They are meant to be the number of non-zero updates for each time, and it looks like maybe they were for some single-key version of the test, and when adding a second key in they were not doubled.

File: src/interchange/src/envelopes.rs (hunk 2 above)
Comment (benesch) [reply src/interchange/src/envelopes.rs:86]: Yeah, that makes sense; the `expect` here is asserting an invariant that is upheld by earlier layers in the system.

**Comment by benesch** [comment]
Thanks, @philip-stoev! I integrated your test into the branch.

File: test/testdrive/github-15748.td [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+# Insert 2 records
+$ kafka-ingest format=avro topic=topic1 schema=${schema}
+{"array":[{"data":{"a":0,"b":0},"time":2,"diff":1}]}
+{"array":[{"data":{"a":15,"b":15},"time":2,"diff":1}]}
+
+# Emit the progress
+$ kafka-ingest format=avro topic=topic1 schema=${schema}
+{"com.materialize.cdc.progress":{"lower":[0],"upper":[10],"counts":[{"time":1,"count":10}, {"time":2,"count":8}]}}
+
+# Because of sort-messages=true, the deletions must be listed first
```
Comment (bkirwi) [inline test/testdrive/github-15748.td:144]: This test doesn't use `sort-messages=true`!

(In general I think we should probably be sorting here... these tests seem like they'd be sensitive to legal changes in the grouping logic or the kafka publishing.)

**Review by bkirwi — COMMENTED** [review]
Overall the new approach here looks good to me, though I am not allowed to formally approve of the branch!

The new tests don't cover what happens if we have multiple keys with the same value, which is something the logic goes to some pains to handle. However, I suspect we probably don't want to specify any particular behaviour here... so that seems fine.

File: test/testdrive/github-15748.td (hunk 3 above)
Comment (philip-stoev) [reply test/testdrive/github-15748.td:144]: Yes sorry the comment needs to be removed

**Comment by bkirwi** [comment]
Yeah... to be clear, it's not obvious to me how to test that behaviour, nor do I think it's worth putting in the effort to do! It's just an oddity of the implementation.

File: test/testdrive/github-15748.td (hunk 3 above)
Comment (benesch) [reply test/testdrive/github-15748.td]: Comment removed! @bkirwi I couldn't figure out how to get `sort-messages` to do what I want—how do you verify that all of ts1 is emitted before all of ts2, for example? So going to merge as is.

**Review by benesch — APPROVED** [review]

**Comment by benesch** [comment]
Thanks, everyone! Merging.

File: test/testdrive/github-15748.td (hunk 3 above)
Comment (bkirwi) [reply test/testdrive/github-15748.td]: As a consumer you're not necessarily guaranteed to see all of `ts1` before `ts2`, depending on how fast you make progress on which partitions. I suppose it's arguable whether you'd rather overspecify or underspecify behaviour for stuff like this, though...

Possibly we should enhance kafka-verify to be able to more precisely state test properties. But we should probably make some headway on figuring out exactly what we want to promise users first...

Comment (benesch) [reply test/testdrive/github-15748.td]: > As a consumer you're not necessarily guaranteed to see all of ts1 before ts2, depending on how fast you make progress on which partitions.

But if it's all one partition we guarantee in-order emission, right?

> Possibly we should enhance kafka-verify to be able to more precisely state test properties. But we should probably make some headway on figuring out exactly what we want to promise users first...

:1000: agree on both counts!

**Comment by philip-stoev** [comment]
Apparently `sort-messages=true` would be needed in both tests :-)

```


testdrive/kafka-avro-upsert-sinks.td:295:1: error: record 4 did not match
--
  | expected:
  | Record {
  | headers: [
  | "3",
  | ],
  | key: Some(
  | {
  | (
  | "a",
  | Long(1),
  | ),
  | },
  | ),
  | value: None,
  | }
  |  
  | actual:
  | Record {
  | headers: [
  | "3",
  | ],
  | key: Some(
  | {
  | (
  | "a",
  | Long(2),
  | ),
  | },
  | ),
  | value: None,
  | }
  | \|

```

I will fix that in a follow-up PR

**Pull request merged**